self_enrollment_tool/views.py

```diff
@@ -186,23 +190,58 @@ def enable(request, course_instance_id):
     roles = literal_eval(request.POST.get('roles'))
     role_id = roles.get('roleId', '')
     role_name = roles.get('roleName', '')
-    logger.debug(f'Selected role_name {role_name} with role_id {role_id} for Self Enrollment in course {course_instance_id}')
+    start_date_str = request.POST.get('start_date', None)
+    end_date_str = request.POST.get('end_date', None)
 
+    start_date = parse_date(start_date_str) if start_date_str else None
+    end_date = parse_date(end_date_str) if end_date_str else None
+
+    eastern = pytz.timezone('US/Eastern')
+
+    if start_date:
+        start_date = eastern.localize(datetime.datetime.combine(start_date, datetime.time.min))
+    if end_date:
+        end_date = eastern.localize(datetime.datetime.combine(end_date, datetime.time.min))
+
+    now_est = timezone.now().astimezone(eastern)
+    today_start_est = eastern.localize(datetime.datetime.combine(now_est.date(), datetime.time.min))
+
+    if start_date and start_date < today_start_est:
+        messages.error(request, "Start date cannot be in the past.")
+        return redirect('self_enrollment_tool:add_new')
+
+    if end_date and end_date < now_est:
+        messages.error(request, "End date must be in the future.")
+        return redirect('self_enrollment_tool:add_new')
+
+    if start_date and end_date and end_date < start_date:
+        messages.error(request, "End date cannot be before the start date.")
+        return redirect('self_enrollment_tool:add_new')
+    
     try:
         if str(role_id) and course_instance_id:
             try:
-                course_instance = SelfEnrollmentCourse.objects.get(course_instance_id=course_instance_id, role_id=role_id)
-                logger.error(f'Self Enrollment is already enabled for this course  {course_instance_id} ')
+                course_instance = SelfEnrollmentCourse.objects.get(
+                    course_instance_id=course_instance_id, 
+                    role_id=role_id, 
+                    start_date__isnull=True,
+                    end_date__isnull=True
+                )
+                logger.info(f'Self Enrollment is already enabled for this course  {course_instance_id} ')
                 messages.error(request, f'Self Enrollment is already enabled for this course (SIS ID {course_instance_id}) and role ({role_name}). '
                               f'See below for the previously-generated link.')
                 path = reverse('self_enrollment_tool:enroll', args=[course_instance.uuid])
```

Comment:
This seems to throw an error which is caught and displayed to the UI as a message in the #263 exception block
`Error creating self enrollment record for course 696751 with role 10. Error details=Reverse for 'enroll' not found. 'enroll' is not a valid view function or pattern name.`

Reply:
![Screen Shot 2024-07-10 at 9 35 44 AM](https://github.com/Harvard-University-iCommons/canvas_account_admin_tools/assets/28874174/fffadfb5-240d-438f-9990-4f1626b5559b)

Reply:
Looking at [URLs](https://github.com/Harvard-University-iCommons/canvas_account_admin_tools/blob/develop/self_enrollment_tool/urls.py) you might want to switch that out for `enable` and possibly put it in the message since the `path` var is not being used anywhere else